### PR TITLE
RFC: Simplify retry behaviour and unify it between plugins

### PR DESCRIPTION
**Pull request by fg91**
RFC for RFC incubator discussion https://github.com/flyteorg/flyte/discussions/3793

### CODE REVIEW

File: rfc/system/0000-simplify-retry-behaviour.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+## 4 Metrics & Dashboards
+
+_NA_
+
+## 5 Drawbacks
+
+Currently the behaviour is intransparent and confusing for platform users. We as platform engineers have to look deep into the code to understand which budget a certain failure counts against. Users cannot increase the number of accepted preemptions when their training runs longer. Different plugins handle preemptions in different ways.
+We should not leave these quirks untackled.
+
+## 6 Alternatives
```
Comment (fellhorn) [inline rfc/system/0000-simplify-retry-behaviour.md:73]: Shall we shorten these sections a little, now that we decided against them?

File: rfc/system/0000-simplify-retry-behaviour.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+## 7 Potential Impact and Dependencies
+
+By having to activate the new retry behaviour in the platform configuration, we will not break existing workflows.
+
+## 8 Unresolved questions
+
+An open question is how we deal with the `interruptibleFailureThreshold`. Currently, this value is a positive integer that is the number of retries that are interruptible. For example, 4 (system failure) retries are allowed where 3 of them are interruptible. Defining it this way does not make as much sense when we allow for task-level configuration of the retries.
+
+We could define the number of retries that ARE NOT interruptible. For example a value of `nonInterruptibleRetries` that is set to 1 means that the last retry is not interruptible regardless of how many retries there are. We also have to discuss whether this value would be configured on the platform side (only for the new retry behaviour) or whether we would add this value to to task-level metadata and expose it to users. Adding this flag while ensuring backwards compatibility will require further discussion.
+
```
Comment (fellhorn) [inline rfc/system/0000-simplify-retry-behaviour.md:113]: Rather something we need to ensure/test: after the change non-retryable exceptions in the user code should not trigger a retry - even when using the PytorchJob plugin.

Probably not worth mentioning? Or do you think this might break?

File: rfc/system/0000-simplify-retry-behaviour.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+The downside of this approach is that the general retry behaviour remains as complex and difficult to understand as it is now.
+
+### 6.2 Demistify failure of Pods belonging to plugins/CRDs
+
+We discussed ways to give plugins access to the status of not only the custom resource they create (e.g. `PyTorchJob`) but also to the statuses of the pods created from the custom resource by the respective third-party operator (e.g. the kubeflow training operator). (This might require changes to the plugin interface, see details [here](https://github.com/flyteorg/flyte/discussions/3793).)
+
+This approach would allow plugins to demistify failures of tasks by looking at the underlying pods which, in contrast to the custom resource, do contain hints of preemptions. This way, we could unify the retry behaviour upon preemptions of the different plugins. However, wouldn't make the retry behaviour easier to understand and users still wouldn't have control over the number of preemptions which is why we opted to not continue these discussions either.
+
+
+### 6.3 Inform the user which retry budget an attempt counted against
```
Comment (fellhorn) [inline rfc/system/0000-simplify-retry-behaviour.md:94]: Maybe remove entirely?

File: rfc/system/0000-simplify-retry-behaviour.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+This approach would allow plugins to demistify failures of tasks by looking at the underlying pods which, in contrast to the custom resource, do contain hints of preemptions. This way, we could unify the retry behavior upon preemptions of the different plugins. However, wouldn't make the retry behavior easier to understand and users still wouldn't have control over the number of preemptions which is why we opted to not continue these discussions either.
+
+
+### 6.3 Inform the user which retry budget an attempt counted against
+
+We considered showing the user in the Flyte console whether an attempt counted against the user retry budget or the system retry budget.
+
+This only improves the transparency and does not remove the different behavior between plugins. Doing so is not necessary when every failure is counted against the same retry budget.
+
+![budgets](https://github.com/flyteorg/flyte/assets/26092524/3e425528-ce3f-42af-8cc3-f0ca0cdab660)
```
Comment (fg91) [inline rfc/system/0000-simplify-retry-behaviour.md:100]:
```suggestion
```

File: rfc/system/0000-simplify-retry-behaviour.md (hunk 2 above)
Comment (fg91) [reply rfc/system/0000-simplify-retry-behaviour.md:108]: https://github.com/flyteorg/flyte/pull/3902/commits/8a9409aee245881aa83c7769e5afede8229ff55f

Good point! I don't think it hurts to be explicit here.

**Comment by davidmirror-ops** [comment]
2023-08-17: Contributor's meeting notes: more discussion is needed to decide if an update to this proposal is needed. There are open questions. Eduardo to ask Union contributors to comment here.

**Comment by fg91** [comment]
@fellhorn and I tried out the WIP implementation of this RFC by @hamersaw on this branch:

https://github.com/flyteorg/flytepropeller/tree/feature/simplify-retries


We ran the following experiments with both the old and the new behaviour:

Test | Old code | New code | Error type | Comment
-- | -- | -- | -- | --
@task(retries=2)<br>def train():<br>     raise Exception("No way back from this") | No retry | No retry | User | Non-retriable exceptions stay non-retriable.
@task(retries=2)<br> def train():<br>    raise FlyteRecoverableException("Please retry”) | 2 retries | 2 retries | User | The handling of recoverable user errors stays the same.
@torch.inference_mode()<br> @task(     retries=2, )<br> def train(): | 2 retries | 2 retries | User | One of our users recently wrapped the task decorator in another decorator which causes the entrypoint to fail. Against expectations this is handled as a user budget retry but the behaviour is not changed.
@torch.inference_mode()<br> @task(retries=2, task_config=Elastic(nnodes=2))<br> def train(): … | 2 retries | 2 retries | User | Same unchanged behaviour e.g. also for the kubeflow plugins
@task(retries=2, interruptible=True)<br> def train():<br>     raise FlyteRecoverableException(”Please retry”) | 2 retries, all interruptible | 2 retries, all interruptible (last retry should be non-interruptible) | User | As explained in the RFC, we think it would be great if also user budget retries switched to non interruptible nodes for the last (n) retry. This is currently not the case but we implemented this functionality in a PR.
@task(retries=1)<br> #  Delete the Pod | 2 retries | 1 retry | System | System retries are now counted against the user defined, unified retry budget as proposed by the RFC.
@task(retries=1,task_config=Elastic())<br> #  Delete the pod | 1 retry | 1 retry | User | With the old behaviour, we see a difference compared to the line above. With the new behaviour, the UX is consistent between the plugin and normal python function task, even though internally in this case we count it as a user retry, not a system retry. It behaves as proposed in the RFC as the user doesn't see a difference.
@task(retries=4, interruptible=True)<br> def train():<br>   raise FlyteRecoverableException(”Please retry”) | 4 retries but all interruptible | # But with our own interruptible logic  4 retries (3 of them interruptible) | User |  


TL;DR:
The WIP implementation removes the inconsistencies in retry behaviour between python function tasks and e.g. the kubeflow plugins and the user is able to control the number of preemptions (simulated by pod deletions) as outlined in the RFC. This makes everything easier to understand.

In @hamersaw's branch, the last retry is not yet switched to non-interruptible as this RFC proposes but we added this functionality this PR:
https://github.com/flyteorg/flytepropeller/pull/610

File: rfc/system/0000-simplify-retry-behaviour.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+This simplified retry behavior does not distinguish between user and system failures but considers a single maximum number of retries (defined by the user in the task decorator) regardless of the cause.
+
+This will require a surprisingly small amount of changes to Flyte:
+
+We propose to:
+
+* Add a configuration flag in FlytePropeller for counting system and user retries the same.
+* Apply the new behavior in the [`isEligibleForRetry`](https://github.com/flyteorg/flytepropeller/blob/294f47a18c9b11892f2d1a3573019e85257c3b19/pkg/controller/nodes/executor.go#L440) function where, in contrast to the current logic, if the flag is set, we check if the number of attempts + the number of system failures is larger than the max number of retries set in the task decorator. For the last retries (details see section 8), non-spot instances are used.
+* Add configuration for default number of retries in FlytePropeller. Currently the default maximum number of attempts is hardcoded. This should be exposed in the configuration.
```
Comment (fg91) [inline rfc/system/0000-simplify-retry-behaviour.md:59]: @hamersaw do we need to remove this sentence? 🤔

File: rfc/system/0000-simplify-retry-behaviour.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+
+
+
+
+## 7 Potential Impact and Dependencies
+
+By having to activate the new retry behavior in the platform configuration, we will not break existing workflows.
+
+## 8 Unresolved questions
```
Comment (fg91) [inline rfc/system/0000-simplify-retry-behaviour.md:101]: TODO:
The implementation we chose in the end needs to be documented here.

**Comment by fg91** [comment]
@davidmirror-ops Is the location proposed in [this](https://github.com/flyteorg/flyte/pull/4022) PR a good place to document that this RFC/behaviour exists?

**Comment by davidmirror-ops** [comment]
> @davidmirror-ops Is the location proposed in [this](https://github.com/flyteorg/flyte/pull/4022) PR a good place to document that this RFC/behaviour exists?

@fg91 Yes, I think it fits well. Thank you!

File: rfc/system/3902-simplify-retry-behaviour.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+## 8 Unresolved questions
+
+An open question is how we deal with the `interruptibleFailureThreshold`. Currently, this value is a positive integer that is the number of retries that are interruptible. For example, 4 (system failure) retries are allowed where 3 of them are interruptible. Defining it this way does not make as much sense when we allow for task-level configuration of the retries.
+
+We could define the number of retries that ARE NOT interruptible. For example a value of `nonInterruptibleRetries` that is set to 1 means that the last retry is not interruptible regardless of how many retries there are. We also have to discuss whether this value would be configured on the platform side (only for the new retry behavior) or whether we would add this value to to task-level metadata and expose it to users. Adding this flag while ensuring backwards compatibility will require further discussion.
+
+Note: while unifying the retry budgets, the new retry behaviour still distinguishes between retriable and non-retriable exceptions and should do so consistently across plugins.
```
Comment (fg91) [inline rfc/system/3902-simplify-retry-behaviour.md:107]:
```suggestion
```

File: rfc/system/3902-simplify-retry-behaviour.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+We propose to introduce a new simplified retry behavior (which will have to be activated in the platform configuration and which will **not** be the default behavior in order to not make a breaking change).
+
+This simplified retry behavior does not distinguish between user and system failures but considers a single maximum number of retries (defined by the user in the task decorator) regardless of the cause.
+
+This will require a surprisingly small amount of changes to Flyte:
+
+We propose to:
+
+* Add a configuration flag in FlytePropeller for counting system and user retries the same.
+* Apply the new behavior in the [`isEligibleForRetry`](https://github.com/flyteorg/flytepropeller/blob/294f47a18c9b11892f2d1a3573019e85257c3b19/pkg/controller/nodes/executor.go#L440) function where, in contrast to the current logic, if the flag is set, we check if the number of attempts + the number of system failures is larger than the max number of retries set in the task decorator. For the last retries (details see section 8), non-spot instances are used.
```
Comment (fg91) [inline rfc/system/3902-simplify-retry-behaviour.md:58]:
```suggestion
* Apply the new behavior in the [`isEligibleForRetry`](https://github.com/flyteorg/flytepropeller/blob/294f47a18c9b11892f2d1a3573019e85257c3b19/pkg/controller/nodes/executor.go#L440) function where, in contrast to the current logic, if the flag is set, we check if the number of attempts + the number of system failures is larger than the max number of retries set in the task decorator. For the last retries, non-spot instances are used.
    * To not introduce an additional platform config flag, the existing flag `interruptibleFailureThreshold` is reused to decide when to switch to a regular instance. With the existing retry behaviour, this number is a positive integer that is the number of retries that are interruptible. For example, 4 (system failure) retries are allowed where 3 of them are interruptible. With the new retry behaviour, a negative number such as e.g. `-2` means that the last to retries are non-interruptible.
```

Comment (fg91) [reply rfc/system/3902-simplify-retry-behaviour.md:58]: @hamersaw do you agree?

File: rfc/system/3902-simplify-retry-behaviour.md (hunk 7 above)
Comment (fg91) [reply rfc/system/3902-simplify-retry-behaviour.md:107]: @hamersaw

**Review by pingsutw — APPROVED** [review]

**Comment by fg91** [comment]
@fellhorn and I opened a PR with some finishing touches into @hamersaw 's PR, see https://github.com/flyteorg/flytepropeller/pull/623. As far as we can see, this completes the implementation of this RFC.

If the others in the steering committee also approve, from my side this is ready @davidmirror-ops .

**Review by pingsutw — APPROVED** [review]

**Review by bstadlbauer — APPROVED** [review]
Looks good to me, thank you!
